giskard/utils/__init__.py

```diff
@@ -109,6 +104,64 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+NB_CANCELLABLE_WORKER_LOCK = Lock()
+
+
+@threaded
+def start_killer(timeout: float, future: Future, pid: int, executor: ProcessPoolExecutor):
+    start = time()
+    # Try to get the result in proper time
+    while (time() - start) < timeout:
+        # future.result(timeout=timeout) => Not working with WSL and python 3.10, switchin to something safer
+        LOGGER.debug("Sleeping for pid %s", pid)
+        sleep(1)
+        if future.done():
+            executor.shutdown(wait=True, cancel_futures=False)
+            with NB_CANCELLABLE_WORKER_LOCK:
+                POOL.nb_cancellable -= 1
+            return
+    LOGGER.warning("Thread gonna kill pid %s", pid)
+    # Manually setting exception, to allow customisation
+    # TODO(Bazire): See if we need a custom error to handle that properly
+    try:
+        future.set_exception(CancelledError("Background task was taking too much time and was cancelled"))
+    except InvalidStateError:
+        pass
+    # Shutting down an executor is actually not stopping the running processes
+    executor.shutdown(wait=False, cancel_futures=False)
+    # Kill the process running by targeting its pid
+    os.kill(pid, signal.SIGINT)
+    # Let's clean up the executor also
+    # Also, does not matter to call shutdown several times
+    executor.shutdown(wait=True, cancel_futures=False)
+    with NB_CANCELLABLE_WORKER_LOCK:
+        POOL.nb_cancellable -= 1
+    LOGGER.debug("Executor has been successfully shutdown")
+    log_pool_stats()
+
+
+def call_in_pool(fn, *args, timeout=None, **kwargs):
+    """Submit the function call with args and kwargs inside the process pool
+
+    Args:
+        fn (function): the function to call
+
+    Returns:
+        Future: the promise of the results
+    """
+    if timeout is None:
+        return POOL.submit(fn, *args, **kwargs)
+    # Create independant process pool
+    # If we kill a running process, it breaking the Process pool, making it unusable
+    one_shot_executor = ProcessPoolExecutor(max_workers=1)
+    pid = one_shot_executor.submit(os.getpid).result(timeout=5)
```

Comment:
I don't get it, the comment above said 
> .result(timeout=timeout) => Not working with WSL and python 3.10

here it's expected to work?

Reply:
I should remove the timeout here anyway, no reason for basic os.getpid to fail or take a long time.

Basically, when I tested it, in the killer thread, the timeout was not respected, and it never stopped, so I had to do the loop.

Btw, the timeout was working out on MacOS with python 3.11.

Reply:
I think the overall killing mechanism could be simplified to avoid having a killer thread and a cancellable counter.

We could use an inter-process data structure to communicate pids between a spawned worker process and the main process: `multiprocessing.Queue` for example.

In this case the first thing the pool process will do is identify its pid and add it to queue, then do the actual work.
In the main process we could use `concurrent.futures._base.as_completed` to wrap the future and add a timeout to it and then call `.add_done_callback(result_handler)` as we do now.

When the timeout is expired we'd find a PID related to a given task and kill it from the main process.

WDYT?

Reply:
> @Hartorn , actually, it looks like https://pypi.org/project/Pebble/ does exactly what we need:
> 
> ```
> from pebble import ProcessPool
> 
> pool = ProcessPool(max_workers=5)
> 
> pool.schedule(long_fn, args=(a,b,c), timeout=1).add_done_callback(done_callback)
> ```

Cancellable counter is only for logs, we can remove it if we want. 
I just wanted to check if that was working, or eventually to avoid having too many process launched.
For the callback of a future, you need everything to be pickable, so you cannot have lock, processPool, and so on.
So I don't know how to shutdown properly the one shot executor we ran.

Also, it cannot be the main executor, since killing a process is breaking the pool, which only raises BrokenPoolException after that.

I'm pretty sure we could avoid the thread is the whole code we were running was async, since a coroutine could do this job, but here I would not be confident.

Reply:
> @Hartorn , actually, it looks like https://pypi.org/project/Pebble/ does exactly what we need:
> 
> ```
> from pebble import ProcessPool
> 
> pool = ProcessPool(max_workers=5)
> 
> pool.schedule(long_fn, args=(a,b,c), timeout=1).add_done_callback(done_callback)
> ```

Looking at it !

Reply:
From what I saw, they are kinda doing the same stuff, when using a separate process, they have a handler thread for watching it and handling timeout.

Should we get this merged and change to use it ? Or want me to switch it ?

Although I'm a bit concerned it's not that much used, the code looks clean

Reply:
Actually, while having a clean API and being easy to use Pebble is a LGPL library so we won't be able to use it (it might be the reason why it's not that widely adopted). 

I also read their code and found similarities with your implementation. I suggest we stick to your current code  (and merge it since we need these changes ASAP).

As an improvement, I think in a separate PR we could inspire from Pebble's API and also encapsulate `call_in_pool` into our custom ProcessPool implementation and have a `schedule` method, WDYT?
In the end Pebble is just a ~5k LOC library

Reply:
Shame, the code was working well with it also.
Sure, seems a good idea to have our own worker implementation. I'm sure it will also help us to improve performance of the lib